File: lv2/12_heap-sorting.py
import heapq

def solution(scoville, K):
    heapq.heapify(scoville)  # 리스트를 최소 힙으로 변환합니다.
    mix_count = 0

    while len(scoville) > 1 and scoville[0] < K:
        first = heapq.heappop(scoville)  # 가장 작은 값
        second = heapq.heappop(scoville)  # 두 번째로 작은 값

        new_scoville = first + second * 2
        heapq.heappush(scoville, new_scoville)  # 새로운 값 삽입

        mix_count += 1

    if scoville[0] < K:
        return -1

    return mix_count

print(solution([1, 2, 3, 9, 10, 12], 7))  # 2

File: sites/programmers/lv2/12_heap-sorting.java
import java.util.PriorityQueue;

public class ScovilleSolution {
    public static int solution(int[] scoville, int K) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int s : scoville) {
            minHeap.offer(s);
        }

        int mixCount = 0;

        while (minHeap.size() > 1 && minHeap.peek() < K) {
            int first = minHeap.poll();
            int second = minHeap.poll();

            int newScoville = first + second * 2;
            minHeap.offer(newScoville);

            mixCount++;
        }

        if (minHeap.peek() < K) {
            return -1;
        }

        return mixCount;
    }

    public static void main(String[] args) {
        int[] scoville = {1, 2, 3, 9, 10, 12};
        int K = 7;
        System.out.println(solution(scoville, K));  // 2
    }
}
